src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -4624,6 +4632,283 @@ std::unique_ptr<LCSerializer> RadosLifecycle::get_serializer(const std::string&
   return std::make_unique<LCRadosSerializer>(store, oid, lock_name, cookie);
 }
 
+RadosRestoreSerializer::RadosRestoreSerializer(RadosStore* store, const std::string& _oid, const std::string& lock_name, const std::string& cookie) :
+  StoreRestoreSerializer(_oid),
+  ioctx(*store->getRados()->get_restore_pool_ctx()),
+  lock(lock_name)
+{
+  lock.set_cookie(cookie);
+}
+
+int RadosRestoreSerializer::try_lock(const DoutPrefixProvider *dpp, utime_t dur, optional_yield y)
+{
+  lock.set_duration(dur);
+  return lock.lock_exclusive((librados::IoCtx*)(&ioctx), oid);
+}
+
+RadosRestore::RadosRestore(RadosStore* _st) : store(_st),
+       	ioctx(*store->getRados()->get_restore_pool_ctx()),
+	r(store->get_neorados()),
+	neo_ioctx(*store->getRados()->get_restore_pool_neo_ctx()) {}
+
+std::unique_ptr<RestoreSerializer> RadosRestore::get_serializer(
+							const std::string& lock_name,
+							const std::string& oid,
+							const std::string& cookie)
+{
+  return std::make_unique<RadosRestoreSerializer>(store, oid, lock_name, cookie);
+}
+
+int RadosRestore::initialize(const DoutPrefixProvider* dpp, optional_yield y,
+		int n_objs, std::vector<std::string>& o_names)
+{
+  num_objs = n_objs;
+  obj_names = o_names;
+
+  for (auto i=0; i < num_objs; i++) {
+    std::unique_ptr<fifo::FIFO> fifo_tmp;
+    try {
+      fifo_tmp = fifo::FIFO::create(dpp, r, obj_names[i], neo_ioctx, ceph::async::use_blocked);
+    } catch (const sys::system_error& e) {
+      ldpp_dout(dpp, -1) << "creating fifo object for index=" << i
+	   << ", objname=" << obj_names[i] << " failed : " << e.what() << dendl;
+      throw;
```

Comment:
This returns an int to indicate errors, so should return an int with `from_error_code` (or `from_exception` if you want to cherry pick that.)

Reply:
Done.